src/Shared/Microsoft.AspNetCore.Razor.Utilities.Shared/PooledObjects/StringHashSetPool.Policy.cs

```diff
@@ -0,0 +1,36 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using Microsoft.Extensions.ObjectPool;
+
+namespace Microsoft.AspNetCore.Razor.PooledObjects;
+
+internal static partial class StringHashSetPool
+{
+    private class Policy : IPooledObjectPolicy<HashSet<string>>
+    {
+        private readonly IEqualityComparer<string>? _comparer;
+
+        public Policy(IEqualityComparer<string>? comparer = null)
+        {
+            _comparer = comparer;
+        }
+
+        public HashSet<string> Create() => new(_comparer);
+
+        public bool Return(HashSet<string> set)
+        {
+            var count = set.Count;
+
+            set.Clear();
+
+            if (count > DefaultPool.MaximumObjectSize)
+            {
+                set.TrimExcess();
```

Comment:
What are your expectations on what this call will do? I hadn't seen this so did a bit of experimentation here to see what would happen. It seems to trim a lot of entries at O(N) cost. I'm wondering is that worth it? Is the O(N) cost here worth it cause at the end of the day we throw away most of the cached allocations here and end up with closer to a `new HashSet<string>()` item. Or should we just return `false` here? 

This is a learning area for me so possible I'm missing a benefit of re-use here in this case. 

This is the experiment I wrote to try to understand better what happens with `TrimExcess` calls. 

```csharp
using System;
using System.Collections.Generic;
using System.Reflection;

var set = new HashSet<string>();
for (int i = 0; i < 100; i++) {
    set.Add(i.ToString());
}
Console.WriteLine(GetCount());
set.Clear();
Console.WriteLine(GetCount());
set.TrimExcess();
Console.WriteLine(GetCount());

int GetCount()
{
    var entries = (System.Array)set
        .GetType()
        .GetField("_entries", BindingFlags.NonPublic | BindingFlags.Instance)
        .GetValue(set)!;
    return entries.Length;
}
```

Outputs

```txt
197
197
3
```

Reply:
My expectation is that it throws away the internal `_buckets` and `_entries` arrays that have grown larger than our threshold. The value is small, but I don't believe it does any additional work for a cleared `HashSet<T>` over creating a new one.

My reading of [`HashSet<T>.TrimExcess()`](https://source.dot.net/#System.Private.CoreLib/src/libraries/System.Private.CoreLib/src/System/Collections/Generic/HashSet.cs,28a57f22e9548ce5) is that it reinitializes the `_buckets` and `_entries` arrays using the size of the next prime after the value of `Count`. Then, it walks through `_entries` up to the value of `Count` (that's the O(n) bit you mentioned). However, because `Count` is 0 after `Clear()`, it doesn't actually walk at all; it just initializes `_buckets` and `_entries` to arrays of length 3.

So, `TrimExcess()` releases `_buckets` and `_entries` to GC. If we return false, `_entries`, `_buckets` *and* the `HashSet<T>` itself will be released. In other words, this does save us a future allocation and doesn't perform any significant extra work after calling `Clear()`. I've included the implementations of `Clear()` and `TrimExcess()` below and annotated them with comments to show my understanding of how it should be working. Please let me know if I've missed something!

```C#
public void Clear()
{
    int count = _count;
    if (count > 0)
    {
        Debug.Assert(_buckets != null, "_buckets should be non-null");
        Debug.Assert(_entries != null, "_entries should be non-null");

        Array.Clear(_buckets);
        _count = 0;
        _freeList = -1;
        _freeCount = 0;
        Array.Clear(_entries, 0, count);
    }
}
```

```C#
public void TrimExcess()
{
    // Set capacity to 0 - the current value of `Count`
    int capacity = Count;

    // Compute the initial size of the `_buckets` and `_entries` arrays based on `Count`. This returns 3.
    int newSize = HashHelpers.GetPrime(capacity);

    // Stick `_entries` in a local variable because we'll likely be accessing it more than once.
    Entry[]? oldEntries = _entries;

    // Get the current length of `_entries`. This is the current capacity of the hash set.
    int currentCapacity = oldEntries == null ? 0 : oldEntries.Length;

    // Is our new size (3) greater than the current capacity? This would only happen for very small hash sets.
    if (newSize >= currentCapacity)
    {
        return;
    }

    // Because we've just cleared the hash set, this sets `oldCount` to 0.
    int oldCount = _count;

    // Update the internal version of the hash set because we reinitialize `_buckets` and `_entries`. (This is used to detect changes during enumeration.)
    _version++;

    // Initialize `_buckets` and `_entries` to new arrays of length 3.
    Initialize(newSize);

    // Stick the new `_entries` in a local variable.
    Entry[]? entries = _entries;
    int count = 0;

    // Walk over `_entries` up to `oldCount` and rebuild the bucket structure. However, because `oldCount` is 0 after `Clear()`, this is just a jump.
    for (int i = 0; i < oldCount; i++)
    {
        int hashCode = oldEntries![i].HashCode;
        if (oldEntries[i].Next >= -1)
        {
            ref Entry entry = ref entries![count];
            entry = oldEntries[i];
            ref int bucket = ref GetBucketRef(hashCode);
            entry.Next = bucket - 1;
            bucket = count + 1;
            count++;
        }
    }

    // Set `_count` and `_freeCount` to 0, which is redundant because `Clear()` already did that.
    _count = capacity;
    _freeCount = 0;
}
```